lib/slack.js

```diff
@@ -49,43 +49,41 @@ class SlackData extends EventEmitter {
       });
   }
 
-  fetch() {
-    request
-      .get(`https://${this.host}.slack.com/api/users.list`)
-      .query({ token: this.token, presence: 1 })
-      .end((err, res) => {
-        this.onres(err, res);
-      });
-    this.emit('fetch');
-  }
-
-  getChannelId(name) {
-    const channel = this.channelsByName[name];
-    return channel ? channel.id : null;
-  }
-
-  retry(delay = this.interval * 2) {
-    setTimeout(this.fetch.bind(this), delay);
-    this.emit('retry');
-  }
-
-  onres(err, res) {
-    if (err) {
-      this.emit('error', err);
-      return this.retry();
-    }
+  async fetch() {
+    let users = [];
+    let firstPage = true;
+    let cursor = '';
 
-    // Too Many Requests
-    if (res.status === 429) {
-      return this.retry(res.headers['retry-after'] * 1000);
-    }
-
-    let users = res.body.members;
+    this.emit('fetch');
 
-    if (!users || !users.length) {
-      this.emit('error', new Error(`Invalid Slack response: ${res.status}`));
-      return this.retry();
+    do {
+      firstPage = false;
+      let response = null;
+      try {
+        response = await request // eslint-disable-line no-await-in-loop
+          .get(`https://${this.host}.slack.com/api/users.list`)
+          .query({
+            token: this.token, limit: 200, cursor, presence: 1,
+          });
+      } catch (err) {
+        this.emit('error', err);
+        return this.retry();
+      }
+      if (response.status === 429) {
+        return this.retry(response.headers['retry-after'] * 1000);
+      }
+      if (response.ok === false) {
+        this.emit('error', new Error(`Slack API error: ${response.error}`));
+        return this.retry();
+      }
+      if (!response.body.members) {
+        this.emit('error', new Error(`Invalid Slack response: ${response.status}`));
+        return this.retry();
+      }
+      users = users.concat(response.body.members);
+      cursor = response.body.response_metadata.next_cursor;
     }
+    while (firstPage === true || cursor !== '');
```

Comment:
Are you sure we need the `firstPage` variable? We should just loop until `cursor !== ''`.

Reply:
Ah good catch. That was a remnant from a less elegant way I've tried first :)